packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1473,6 +1485,144 @@ class _SelectableFragment with Selectable, ChangeNotifier implements TextLayoutM
+    final TextPosition? existingSelectionEnd = _textSelectionEnd;
+
+    _setSelectionPosition(null, isEnd: isEnd);
+    final Matrix4 transform = paragraph.getTransformTo(null);
+    transform.invert();
+    final Offset localPosition = MatrixUtils.transformPoint(transform, globalPosition);
+    if (_rect.isEmpty) {
+      return SelectionUtils.getResultBasedOnRect(_rect, localPosition);
+    }
+    final Offset adjustedOffset = SelectionUtils.adjustDragOffset(
+      _rect,
+      localPosition,
+      direction: paragraph.textDirection,
+    );
+
+    final TextPosition position = paragraph.getPositionForOffset(adjustedOffset);
+    // Check if the original local position is within the rect, if it is not then
+    // we do not need to look up the word boundary for that position. This is to
+    // maintain a selectables selection collapsed at 0 when the local position is
+    // not located inside its rect.
+    final ({TextPosition wordStart, TextPosition wordEnd})? wordBoundary = !_rect.contains(localPosition) ? null : _getWordBoundaryAtPosition(position);
+    TextPosition? targetPosition;
+    if (wordBoundary != null) {
+      assert(wordBoundary.wordStart.offset >= range.start && wordBoundary.wordEnd.offset <= range.end);
+      if (existingSelectionStart != null && existingSelectionEnd != null) {
+        final TextPosition edgeToKeepInPlace = isEnd ? existingSelectionStart : existingSelectionEnd;
+        final TextPosition currentMovingEdge = isEnd ? existingSelectionEnd : existingSelectionStart;
+        if (_selectableContainsOriginWord) {
+          // Swap the selection edges to keep the origin word in bounds, when the
+          // new position where to cause an inverted selection.
+          final bool shouldSwapEdgesWhenSelectionIsNormalized = isEnd ? position.offset < existingSelectionStart.offset : position.offset > existingSelectionEnd.offset;
+          final bool shouldSwapEdgesWhenSelectionNotNormalized = isEnd ? position.offset > existingSelectionStart.offset : position.offset < existingSelectionEnd.offset;
+          if (existingSelectionStart.offset <= existingSelectionEnd.offset) {
+            // The selection is normalized, i.e. the start of the selection is before
+            // the end.
+            if (shouldSwapEdgesWhenSelectionIsNormalized) {
+              targetPosition = isEnd ? wordBoundary.wordStart : wordBoundary.wordEnd;
+              // The static edge is moved to the opposite end of the word boundary to
+              // keep the entire origin word in the selection.
+              final ({TextPosition wordStart, TextPosition wordEnd}) localWordBoundary = _getWordBoundaryAtPosition(edgeToKeepInPlace);
+              assert(localWordBoundary.wordStart.offset >= range.start && localWordBoundary.wordEnd.offset <= range.end);
+              _setSelectionPosition(isEnd ? localWordBoundary.wordEnd : localWordBoundary.wordStart, isEnd: !isEnd);
```

Comment:
Nit: Is this any simpler?

```dart
final (wordStart: TextPosition wordStart, wordEnd: TextPosition wordEnd) = _getWordBoundaryAtPosition(edgeToKeepInPlace);
```

So you don't have to refer to localWordBoundary.  Here and other places where you receive a record.